File: 백준/Silver/25918. 북극곰은 괄호를 찢어/북극곰은 괄호를 찢어.cc
#include <iostream>
#include <stack>
#define endl '\n'
using namespace std;

int main()
{
    int N;
    cin >> N;

    stack<char> s;

    int day = 0;
    for (int i = 0; i < N; i++)
    {
        char a;
        cin >> a;

        if (s.empty() || s.top() == a)
        {
            s.push(a);
        }
        else if (!s.empty() || s.top() != a)
        {
            s.pop();
        }

        if (s.size() > day)
        {
            day = s.size();
        }
    }

    if (!s.empty())
    {
        cout << -1 << endl;
    }
    else
    {
        cout << day << endl;
    }
}

File: 백준/Silver/25918. 북극곰은 괄호를 찢어/북극곰은 괄호를 찢어.py
from collections import deque

N = int(input())

s = deque()
a = input()

day = 0
for i in range(len(a)):
    if not s or s[-1] == a[i]:
        s.append(a[i])
    else:
        s.pop()

    if len(s) > day:
        day = len(s)

if s:
    print(-1)
else:
    print(day)
